### PR TITLE
UPSTREAM: <carry>: apiextensions: make CR registry reusable with different store

### PR DESCRIPTION
**Pull request by sttts**

### CODE REVIEW

**Review by davidfestal — APPROVED** [review]

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -663,7 +663,7 @@ func (r *crdHandler) tearDown(oldInfo *crdInfo) {
 
 	for _, storage := range oldInfo.storages {
 		// destroy only the main storage. Those for the subresources share cacher and etcd clients.
-		storage.CustomResource.DestroyFunc()
+		storage.CustomResource.Store.(*genericregistry.Store).DestroyFunc()
```
Comment (ncdc) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:666]: Could we put DestroyFunc() on the new Store interface?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_patch.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (ncdc) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_patch.go:2]: I've been doing kcp for our patch files

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_patch.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+package customresource
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/registry/generic"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+)
+
+type Store interface {
```
Comment (ncdc) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_patch.go:26]: This file needs godoc

File: staging/src/k8s.io/apiserver/pkg/authentication/group/authenticated_group_adder.go [hunk 4]
```diff
@@ -51,7 +51,8 @@ func (g *AuthenticatedGroupAdder) AuthenticateRequest(req *http.Request) (*authe
 		}
 	}
 
-	r.User = &user.DefaultInfo{
+	ret := *r // shallow copy
```
Comment (ncdc) [inline staging/src/k8s.io/apiserver/pkg/authentication/group/authenticated_group_adder.go:58]: Why / how is this related to the store changes?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_patch.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd_patch.go:2]: ack. Was fearing some verify script will get unhappy.

File: staging/src/k8s.io/apiserver/pkg/authentication/group/authenticated_group_adder.go (hunk 4 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/authentication/group/authenticated_group_adder.go:58]: not sure how I triggered it, but it broke in my e2es. Upstream merged. So we can drop in next rebase.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:666]: https://github.com/kcp-dev/kubernetes/pull/68/files
